### PR TITLE
Improvements to Materializations ElasticSearch

### PR DESCRIPTION
**Pull request by jixij**
The PR improved the materialization elastic search connector.
1. it exports the descriptions of the connect specifications in the spec responses.
2. it adds integration tests to it. As the ingest to flow mechanism has been changed recently, the change also involves changes to the materialization testing infrastructure, and the existing integration test for `s3-parquet.`

The changes are structured roughly in a sequence of commits.
- `data ingest script` adds a new script to ingest data to Flow captures via `Push` mechnism.
- `test infra change and s3-parquet` changes the testing infrastructure to use the new ingest script and switched the existing test to work with the new infra.
- `elasticsearch tests` adds the integration test for elastic search.
- `hook with github workflow` enables the integration tests to be triggered in github workflow.
- `change elastic search jsonschema descriptions` implements requirement 1.

### CODE REVIEW

**Comment by jixij** [comment]
sending out for a review
`source-s3` integration testing is failing for now. And it should not be caused by this PR. And instead, most likely cause by aws infra issue that makes the [bucket creation operation to fail](https://github.com/estuary/connectors/runs/4448744293?check_suite_focus=true#step:14:68).

From https://status.aws.amazon.com/
```
[11:26 AM PST] We are seeing impact to multiple AWS APIs in the US-EAST-1 Region. This issue is also affecting some of our monitoring and incident response tooling, which is delaying our ability to provide updates. Services impacted include: EC2, Connect, DynamoDB, Glue, Athena, Timestream, and Chime and other AWS Services in US-EAST-1. The root cause of this issue is an impairment of several network devices in the US-EAST-1 Region. We are pursuing multiple mitigation paths in parallel, and have seen some signs of recovery, but we do not have an ETA for full recovery at this time. Root logins for consoles in all AWS regions are affected by this issue, however customers can login to consoles other than US-EAST-1 by using an IAM role for authentication.
```

File: tests/materialize/datasets/ingest.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+		return fmt.Errorf("open file %s: %w", dataFilePath, err)
+	}
+	var scanner = bufio.NewScanner(file)
+	scanner.Split(bufio.ScanLines)
+
+	var arena flow.Arena
+	var docs []flow.Slice
+	for scanner.Scan() {
+		docs = append(docs, arena.Add(scanner.Bytes()))
+	}
```
Comment (psFried) [inline tests/materialize/datasets/ingest.go:73]: need to check `scanner.Err()` here?

File: tests/materialize/materialize-s3-parquet/ingest-data.sh [hunk 2]
```diff
@@ -1,16 +1,22 @@
...
 
-pv --line-mode --quiet --rate-limit 1 < "${DATASET_MULTIPLE_DATATYPES}" \
-      | websocat --protocol json/v1 "${WEBSOCKET_URL_MULTIPLE_DATATYPES}"
+go run ${DATA_INGEST_SCRIPT_PATH} \
+  --server_address=${CONSUMER_ADDRESS} \
+  --capture=${PUSH_CAPTURE_NAME} \
+  --binding_num=${BINDING_NUM_MULTIPLE_DATATYPES} \
+  --data_file_path=${DATASET_MULTIPLE_DATATYPES}
 
 sleep 5
```
Comment (psFried) [inline tests/materialize/materialize-s3-parquet/ingest-data.sh:17]: nit: Add a comment? I don't think it's obvious why this `sleep` is here without one

File: tests/materialize/run.sh [hunk 3]
```diff
@@ -2,116 +2,134 @@
...
         --mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock \
-        --mount type=bind,source="$(pwd)",target=/home/flow/project \
+        --mount type=bind,source="${TEST_SCRIPTS_DIR}",target=${test_scripts_dir_target} \
+        --mount type=bind,source="${TEST_DIR}",target=${test_dir_target} \
+        --env BROKER_ADDRESS="http://${BROKER_ADDRESS}" \
+        --env CONSUMER_ADDRESS="http://${CONSUMER_ADDRESS}" \
+        --env TEST_DIR="${test_dir_target}" \
+        --env BUILDS_ROOT="file://${test_dir_target}/build/" \
+        --env BUILD_ID=run-test-"${CONNECTOR}" \
+        --env CATALOG \
```
Comment (psFried) [inline tests/materialize/run.sh:103]: What is the value of `CATALOG` here? Is that taking the value that's currently set in the environment?

**Review by psFried — APPROVED** [review]
Wow, thanks for getting all the tests running! Overall this looks very good. I had a few minor things to say, but I think it's ok to merge after addressing those.

File: materialize-elasticsearch/driver.go [hunk 4]
```diff
@@ -55,6 +51,21 @@ func (r resource) Validate() error {
 	return nil
 }
 
+func (resource) GetFieldDocString(fieldName string) string {
```
[the remark this answers is not in the document]
Comment (jixij) [reply materialize-elasticsearch/driver.go:55]: Done.

File: tests/materialize/materialize-s3-parquet/ingest-data.sh (hunk 2 above)
Comment (jixij) [reply tests/materialize/materialize-s3-parquet/ingest-data.sh:19]: Done.

File: tests/materialize/run.sh (hunk 3 above)
Comment (jixij) [reply tests/materialize/run.sh:103]: Yes, exactly. This is an abbreviation of `--env CATALOG=${CATALOG}`, that docker run recognize.

**Comment by jixij** [comment]
Cool. Thank you Phil for the review!

**Review by jgraettinger — CHANGES_REQUESTED** [review]
Agreed, neat work getting materialization tests running with the ingest RPC. Really cool to see.

I have some comments below on the test running mechanisms.

> Reviewed 19 of 23 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/estuary/connectors/80#-:-MqUbzq45fbnfcQtDYjr:b-5upty0)* status: all files reviewed, 8 unresolved discussions (waiting on @jixij and @saterus)

___
*[tests/materialize/run.sh, line 1 at r2](https://reviewable.io/reviews/estuary/connectors/80#-MqUXSdQ-M7TTxfbPXjL:-MqUXSdQ-M7TTxfbPXjM:b1rxnjv) ([raw file](https://github.com/estuary/connectors/blob/3c7be641f93436da50c814441f027b0c18f611fb/tests/materialize/run.sh#L1)):*
> ```Shell
> #!/bin/bash
> ```

Check out: https://github.com/estuary/connectors/blob/main/tests/run.sh

There are some big differences in how these two scripts are working, that we should refactor to share common behavior or at least standardize across them. Differences I'm noting:

* Running inside docker or outside. A bit back I factored `run.sh` to use docker images for delivery, but copy out binaries and run them _outside_ of docker. IMO running inside containers makes things more complicated and harder to debug.
* Using Unix sockets instead of port bindings for data-plane components (BROKER_ADDRESS, CONSUMER_ADDRESS). This allows parallel tests to run in isolation on the same machine without clobbering each other.
___
*[tests/materialize/flowctl/build-and-activate.sh, line 11 at r2](https://reviewable.io/reviews/estuary/connectors/80#-MqU_5QxDJGxMLyb4Qhp:-MqU_5QxDJGxMLyb4Qhq:b-kkkctr) ([raw file](https://github.com/estuary/connectors/blob/3c7be641f93436da50c814441f027b0c18f611fb/tests/materialize/flowctl/build-and-activate.sh#L11)):*
> ```Shell
>   --build-id "${BUILD_ID}" \
>   --source "file://${TEST_DIR}/${CATALOG}" \
>   --log.level debug
> ```

nit: debug log level is perhaps a bit high ?
___
*[tests/materialize/flowctl/temp-data-plane.sh, line 8 at r2](https://reviewable.io/reviews/estuary/connectors/80#-MqU_rh89wPw-cIYJ4QV:-MqU_rh89wPw-cIYJ4QW:b-rmk4o1) ([raw file](https://github.com/estuary/connectors/blob/3c7be641f93436da50c814441f027b0c18f611fb/tests/materialize/flowctl/temp-data-plane.sh#L8)):*
> ```Shell
> 
> flowctl temp-data-plane \
>     --sigterm \
> ```

The argument for _not_ using --sigterm is that we want to know if we're failing to clean up resources on test exit.

What do you think about in-lining these scripts running `flowctl` ? They add indirection for the code reader, but I'm not sure they're encapsulating enough functionality to really be worth it.
___
*[tests/materialize/materialize-s3-parquet/setup.sh, line 17 at r2](https://reviewable.io/reviews/estuary/connectors/80#-MqUbGke-6byIVRZVuks:-MqUbGke-6byIVRZVukt:b-csq1rf) ([raw file](https://github.com/estuary/connectors/blob/3c7be641f93436da50c814441f027b0c18f611fb/tests/materialize/materialize-s3-parquet/setup.sh#L17)):*
> ```Shell
> export TEST_PATH_PREFIX_MULTIPLE_DATATYPES="${CONNECTOR}/multiple-datatypes"
> 
> function startLocalStack() {
> ```

I'm unsure about adding localstack, because it puts us in an uncomfortable position where we're straddling a line of _some_ connectors using it, and _others_ not. It's a new dependency, a new thing to maintain and for everybody to have to know about and understand.

We've used it before, and the reason we took it out is because localstack is emulated: It's subtly different than the real services, and you _still_ need to test against the real thing to be totally sure. And it simply wasn't any harder to test against the real thing.

This is especially true with `sops` and easy dev access to credentials required for testing. We have some work to do still in this repo to make it as easy as it could be (rather than plumbing through git secrets), but it's achievable.

**Comment by jixij** [comment]
> Agreed, neat work getting materialization tests running with the ingest RPC. Really cool to see.
> 
> I have some comments below on the test running mechanisms.
> 
> > Reviewed 19 of 23 files at r1, 4 of 4 files at r2, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/estuary/connectors/80#-:-MqUbzq45fbnfcQtDYjr:b-5upty0)_ status: all files reviewed, 8 unresolved discussions (waiting on @jixij and @saterus)
> 
> _[tests/materialize/run.sh, line 1 at r2](https://reviewable.io/reviews/estuary/connectors/80#-MqUXSdQ-M7TTxfbPXjL:-MqUXSdQ-M7TTxfbPXjM:b1rxnjv) ([raw file](https://github.com/estuary/connectors/blob/3c7be641f93436da50c814441f027b0c18f611fb/tests/materialize/run.sh#L1)):_
> 
> > ```shell
> > #!/bin/bash
> > ```
> 
> Check out: https://github.com/estuary/connectors/blob/main/tests/run.sh
> 
> There are some big differences in how these two scripts are working, that we should refactor to share common behavior or at least standardize across them. Differences I'm noting:
> 
> * Running inside docker or outside. A bit back I factored `run.sh` to use docker images for delivery, but copy out binaries and run them _outside_ of docker. IMO running inside containers makes things more complicated and harder to debug.
> * Using Unix sockets instead of port bindings for data-plane components (BROKER_ADDRESS, CONSUMER_ADDRESS). This allows parallel tests to run in isolation on the same machine without clobbering each other.
> 
> _[tests/materialize/flowctl/build-and-activate.sh, line 11 at r2](https://reviewable.io/reviews/estuary/connectors/80#-MqU_5QxDJGxMLyb4Qhp:-MqU_5QxDJGxMLyb4Qhq:b-kkkctr) ([raw file](https://github.com/estuary/connectors/blob/3c7be641f93436da50c814441f027b0c18f611fb/tests/materialize/flowctl/build-and-activate.sh#L11)):_
> 
> > ```shell
> >   --build-id "${BUILD_ID}" \
> >   --source "file://${TEST_DIR}/${CATALOG}" \
> >   --log.level debug
> > ```
> 
> nit: debug log level is perhaps a bit high ?
> 
> _[tests/materialize/flowctl/temp-data-plane.sh, line 8 at r2](https://reviewable.io/reviews/estuary/connectors/80#-MqU_rh89wPw-cIYJ4QV:-MqU_rh89wPw-cIYJ4QW:b-rmk4o1) ([raw file](https://github.com/estuary/connectors/blob/3c7be641f93436da50c814441f027b0c18f611fb/tests/materialize/flowctl/temp-data-plane.sh#L8)):_
> 
> > ```shell
> > flowctl temp-data-plane \
> >     --sigterm \
> > ```
> 
> The argument for _not_ using --sigterm is that we want to know if we're failing to clean up resources on test exit.
> 
> What do you think about in-lining these scripts running `flowctl` ? They add indirection for the code reader, but I'm not sure they're encapsulating enough functionality to really be worth it.
> 
> _[tests/materialize/materialize-s3-parquet/setup.sh, line 17 at r2](https://reviewable.io/reviews/estuary/connectors/80#-MqUbGke-6byIVRZVuks:-MqUbGke-6byIVRZVukt:b-csq1rf) ([raw file](https://github.com/estuary/connectors/blob/3c7be641f93436da50c814441f027b0c18f611fb/tests/materialize/materialize-s3-parquet/setup.sh#L17)):_
> 
> > ```shell
> > export TEST_PATH_PREFIX_MULTIPLE_DATATYPES="${CONNECTOR}/multiple-datatypes"
> > 
> > function startLocalStack() {
> > ```
> 
> I'm unsure about adding localstack, because it puts us in an uncomfortable position where we're straddling a line of _some_ connectors using it, and _others_ not. It's a new dependency, a new thing to maintain and for everybody to have to know about and understand.
> 
> We've used it before, and the reason we took it out is because localstack is emulated: It's subtly different than the real services, and you _still_ need to test against the real thing to be totally sure. And it simply wasn't any harder to test against the real thing.
> 
> This is especially true with `sops` and easy dev access to credentials required for testing. We have some work to do still in this repo to make it as easy as it could be (rather than plumbing through git secrets), but it's achievable.

Got it. Thank you Johnny, I will keep working based on your feedback.

**Comment by jixij** [comment]
___
*[tests/materialize/run.sh, line 1 at r2](https://reviewable.io/reviews/estuary/connectors/80#-MqUXSdQ-M7TTxfbPXjL:-Mqo95sg3cNNTirEg8E9:btkp63s) ([raw file](https://github.com/estuary/connectors/blob/3c7be641f93436da50c814441f027b0c18f611fb/tests/materialize/run.sh#L1)):*
<details><summary><i>Previously, jgraettinger (Johnny Graettinger) wrote…</i></summary><blockquote>

Check out: https://github.com/estuary/connectors/blob/main/tests/run.sh

There are some big differences in how these two scripts are working, that we should refactor to share common behavior or at least standardize across them. Differences I'm noting:

* Running inside docker or outside. A bit back I factored `run.sh` to use docker images for delivery, but copy out binaries and run them _outside_ of docker. IMO running inside containers makes things more complicated and harder to debug.
* Using Unix sockets instead of port bindings for data-plane components (BROKER_ADDRESS, CONSUMER_ADDRESS). This allows parallel tests to run in isolation on the same machine without clobbering each other.
</blockquote></details>

I am working on the PR https://github.com/estuary/connectors/pull/84/files to refactor the test code to run outside docker, so that both source and materialize connector scripts are in sync. That simplified things a lot, and it allows to enable the unix socket, which was not easy to structure nicely when running inside docker due to the additional requirements on permission settings of the socket file.

One thoughts while working on it - there are three places that needs the testing scripts (E2E test in Flow + Source/Materialize Tests in Connectors). In the future, maybe the common testing scripts could live somewhere in Flow Repo, and shared by all three places. (The connector tests could also copy them out from Flow Image as well.)

I'll look into the S3 changes later on, and will incorporate that into the PR as well. Thanks!

**Pull request merged**